ravenml/train/commands.py

```diff
@@ -8,93 +8,84 @@
+        ctx.obj = TrainInput(train_config, ctx.invoked_subcommand)
 
 @train.resultcallback()
+@pass_train
 @click.pass_context
-# def process_result(ctx: click.Context, result: TrainOutput, local: str, dataset: str):
-def process_result(ctx: click.Context, result: TrainOutput, local: str, dataset: str,  no_kill: bool):
+def process_result(ctx: click.Context, ti: TrainInput, result: TrainOutput, config: str):
     """Processes the result of a training by analyzing the given TrainOutput object.
     This callback is called after ANY command originating from the train command 
-    group, hence the check for commands other than training plugins.
+    group, hence the check to see if a result was actually returned - plugins
+    simply do not return a TrainOutput from non-training commands.
 
     Args:
         ctx (Context): click context object
+        ti (TrainInput): TrainInput object stored at ctx.obj, created at start of training.
+            This object contains the metadata from the run and the path to the
+            training artifacts (at ti.artifact_path) which is printed to the user
+            after a local training.
         result (TrainOutput): training output object returned by training plugin
-        local (str): copy of local option provided to original command (see train)
-        dataset (str): copy of dataset option provided to original comamand (see train)
+        config (str): config option from train command. Click requires that command
+            callbacks accept the options from the original command.
     """
-
-    # need to consider issues with this being called on every call to train
-    if ctx.invoked_subcommand != 'list' and result is not None:
-        # upload if not in local mode
-        if not result.local_mode:
-            uuid = cli_spinner('Uploading artifacts...', _upload_result, result)
+    if result is not None:
+        # upload if not in local mode, determined by user defined artifact_path field in config
+        if not ti.config.get('artifact_path'):
+            uuid = cli_spinner('Uploading artifacts...', _upload_result, result, ti.metadata)
             click.echo(f'Artifact UUID: {uuid}')
         else:
-            click.echo(f'LOCAL MODE: Not uploading model to S3. Model is located at: {result.artifact_path}')
+            with open(ti.artifact_path / 'metadata.json', 'w') as f:
+                json.dump(ti.metadata, f, indent=2)
+            click.echo(f'LOCAL MODE: Not uploading model to S3. Model is located at: {ti.artifact_path}')
             
-        # kill if on ec2
-        if not no_kill:
-            click.echo('Attempting to kill EC2 instance...')
+        # kill if on ec2 unless user has explicitly said not to
+        # NOTE: users may run in local mode on EC2. Do not fall into the trap of trying
+        # to use the artifact_path field in the config to determine if running on EC2 or not.
+        if not ti.config.get('ec2_no_kill'):
```

Comment:
This is more of a feature request, so it doesn't have to make it into this PR if you don't have the time to implement it: but instead of `ec2_no_kill`, can we have `ec2_kill: 'stop' | 'terminate' | 'none`, and have it default to `stop`? This would align a lot more with how we currently use EC2 instances.

Reply:
Adding now, default if field is not provided in config is `stop`. Users can also explicitly set `stop`, or set to `terminate`. Any string other than `stop` or `terminate` results in keeping the instance alive (so `none` would work as well)